llvm/lib/CAS/HashMappedTrie.cpp

```diff
@@ -445,11 +431,12 @@ ThreadSafeHashMappedTrieBase::ThreadSafeHashMappedTrieBase(
 }
 
 ThreadSafeHashMappedTrieBase::~ThreadSafeHashMappedTrieBase() {
-  assert(!ImplPtr && "Expected subclass to call destroyImpl()");
+  assert(!ImplPtr.load() && "Expected subclass to call destroyImpl()");
 }
 
-void ThreadSafeHashMappedTrieBase::destroyImpl(function_ref<void (void *)> Destructor) {
-  std::unique_ptr<ImplType> Impl = ImplPtr.take();
+void ThreadSafeHashMappedTrieBase::destroyImpl(
+    function_ref<void(void *)> Destructor) {
+  ImplType *Impl = ImplPtr.load();
```

Comment:
Seems like this should be an `exchange(nullptr)` so that if there is another thread, it'll see `nullptr` here. Specifically, I suggest:
```
std::unique_ptr<ImplType> Impl(ImplPtr.exchange(nullptr));
```
This is equivalent to the previous `ImplPtr.take()`.